src/engine/sparqlExpressions/ConvertToDtypeConstructor.cpp

```diff
@@ -103,8 +124,55 @@ class ToBooleanImpl {
 using ToBoolean = NARY<1, FV<ToBooleanImpl, ToNumericValueGetter>>;
 }  // namespace detail::to_boolean
 
+//______________________________________________________________________________
+// CONVERT TO DATE(TIME)
+namespace detail::to_datetime {
+
+// Cast to xsd:dateTime or xsd:date (ValueId)
+template <bool ToJustXsdDate>
+inline auto convertStringToDateTimeValueId =
+    [](OptIdOrString input) -> ValueId {
+  if (!input.has_value()) {
+    return Id::makeUndefined();
+  }
+  const auto& inputValue = input.value();
+
+  // Remark: If the parsing procedure for datetime/date string values with
+  // parseXsdDatetimeGetOptDate/parseXsdDateGetOptDate fails,
+  // return Id::makeUndefined().
+  const auto retrieveValueId = [](std::optional<DateYearOrDuration> optValue) {
+    if (optValue.has_value()) {
+      return Id::makeFromDate(optValue.value());
+    }
+    return Id::makeUndefined();
+  };
+
+  if (auto* valueId = std::get_if<ValueId>(&inputValue)) {
+    return *valueId;
+  }
```

Comment:
Do this first, before the definition of the `retrieveValueId`, includes readability.
And consider using `std::visit()` which is more typesafe against future changes instead of the `if(get_if)` chains.